### ISSUE TITLE
Decoding fails when only providing a decoder for the top level class in Scala 3

### ISSUE THREAD

**Issue by bjornakr** [circe/circe#2001]
In the following snippet I have only derived a decoder for the parent class B. When trying to use the decoder, it gives a DecodingFailure on runtime, unless I also provide a decoder for the child class B. This only seems to be a problem when the child class is optional. I would expect either a compilation error like you would get in Scala 2, or a working decoder for the whole structure.

Any ideas why this happens?

```scala
import io.circe.Decoder
import io.circe.syntax._
import io.circe.parser.decode
import io.circe.generic.semiauto.deriveDecoder

case class A(i: Int)
case class B(a: Option[A])

// given Decoder[A] = deriveDecoder       // With this line, decoding succeeds
given Decoder[B] = deriveDecoder

println(decode[B]("""{"a": {"i": 1}}""")) // Left(DecodingFailure(Option, List(DownField(a))))
println(decode[B]("""{"a": null}"""))     // Left(DecodingFailure(Option, List(DownField(a))))
println(decode[B]("""{}"""))              // Left(DecodingFailure(Missing required field, List(DownField(a))))
```

Try it out here: https://scastie.scala-lang.org/hZ2gjH6IQnOV4ZMAQN7bUg

Also, thank you for this great library!

**Comment by armanbilge** [comment]
`case class B(a: Option[A]) derives Decoder` seems to work. This is the idiomatic way to write it in Scala 3.

**Comment by matsluni** [comment]
Hi @bjornakr , I stumbled upon this ticket...Can this be closed, now that #2258 is merged?

I tried the scastie with circe version 0.14.9, scala 3.3.0 (and above) and I now get a compile error if the decoder for `A` is commented.

**Comment by bjornakr** [comment]
Yes, I double checked, and I get the same result. Thanks for looking into it. Closing this now.
